### ISSUE TITLE
Neural Network Bot Runner

### ISSUE THREAD

**Issue by tgrunnagle** [tgrunnagle/go-towerfall#96]
## Description

Create `NeuralNetBot` class that uses `ActorCriticNetwork` for inference and `NeuralNetBotRunner` that integrates with `GameClient` to enable trained neural network models to play in end-user games. This is the foundational component that allows ML-trained bots to make real-time decisions based on game state.

## Context

This task is the root task for the "Add Bots to End-User Games" epic. It establishes the core inference pattern for neural network bots, following the existing `RuleBasedBot` and `RuleBasedBotRunner` patterns in the codebase. The `NeuralNetBot` will convert game state to observations, run inference through a trained `ActorCriticNetwork`, and translate discrete actions back to game inputs.

**Parent Issue**: #95
**Dependencies**: None (root task)
**Blocks**: TASK-002 (WebSocket Bot Client), TASK-003 (Bot Manager)

## Acceptance Criteria

- [ ] `NeuralNetBot` class extends `BaseBot` and implements `decide_actions()`
- [ ] `NeuralNetBot` uses `ObservationBuilder` to convert `GameState` to observation vectors
- [ ] `NeuralNetBot` runs inference through `ActorCriticNetwork.get_action_and_value()` with deterministic=True
- [ ] `NeuralNetBot` translates discrete `Action` enum values to `BotAction` tuples (keyboard/mouse actions)
- [ ] `NeuralNetBotRunner` follows the same pattern as `RuleBasedBotRunner`
- [ ] `NeuralNetBotRunner` tracks previous action state to send only changed inputs
- [ ] `NeuralNetBotRunner` has `reset()` method for new game/episode
- [ ] `NeuralNetBotConfig` dataclass for configuration options
- [ ] Module exports added to `bot/bots/__init__.py`
- [ ] All unit tests pass (`task bot2:test:unit`)
- [ ] Code passes linting and type checking (`task bot2:check`)

### Recommended Implementation

Follow the existing `RuleBasedBot` and `RuleBasedBotRunner` patterns. The key difference is that `NeuralNetBot` uses neural network inference instead of heuristic rules:

1. **Observation Building**: Use `ObservationBuilder` to convert `GameState` to a normalized numpy array
2. **Inference**: Convert observation to PyTorch tensor, run through `ActorCriticNetwork.get_action_and_value(deterministic=True)`
3. **Action Translation**: Use the `Action` enum and helper functions from `bot/actions.py` to convert discrete actions to game inputs

### Patterns & Frameworks

- Follow `RuleBasedBot` pattern for `BaseBot` extension
- Follow `RuleBasedBotRunner` pattern for `GameClient` integration
- Use `ObservationBuilder` from `bot.observation.observation_space`
- Use `Action` enum and `execute_action()` pattern from `bot.actions`
- Use Pydantic v2 dataclass patterns for configuration
- Use native Python typing (e.g., `list` not `typing.List`, `str | None` not `Optional[str]`)

### Code Pointers

- `bot2/src/bot/bots/rule_based_bot.py` - Reference for bot and runner implementation patterns (lines 47-442)
- `bot2/src/bot/bots/base_bot.py` - Base class to extend, defines `BotAction` types (lines 20-81)
- `bot2/src/bot/agent/network.py` - `ActorCriticNetwork` for inference (lines 17-211)
- `bot2/src/bot/observation/observation_space.py` - `ObservationBuilder` for game state conversion (lines 98-432)
- `bot2/src/bot/actions.py` - `Action` enum and action execution helpers (lines 31-289)
- `bot2/src/bot/bots/__init__.py` - Module exports to update (lines 1-44)
- `bot2/tests/unit/test_rule_based_bot.py` - Test patterns for bot tests
- `bot2/tests/unit/test_network.py` - Test patterns for network-related tests

### Component Interfaces

```python
# bot2/src/bot/bots/neural_net_bot.py

from dataclasses import dataclass
import torch
from bot.agent.network import ActorCriticNetwork
from bot.bots.base_bot import BaseBot, BotAction
from bot.client import GameClient
from bot.models import GameState
from bot.observation.observation_space import ObservationBuilder, ObservationConfig

@dataclass
class NeuralNetBotConfig:
    """Configuration for neural network bot behavior."""
    observation_config: ObservationConfig | None = None
    device: str = "cpu"  # Device for inference ("cpu" or "cuda")


class NeuralNetBot(BaseBot):
    """Neural network bot that uses ActorCriticNetwork for inference.

    Uses a trained ActorCriticNetwork to make decisions based on game state.
    The bot converts game state to observations, runs inference, and translates
    discrete actions to game inputs.
    """

    def __init__(
        self,
        player_id: str,
        network: ActorCriticNetwork,
        config: NeuralNetBotConfig | None = None,
    ) -> None:
        """Initialize the neural network bot.

        Args:
            player_id: Unique identifier for this bot's player
            network: Trained ActorCriticNetwork for inference
            config: Optional configuration for tuning bot behavior
        """
        ...

    async def decide_actions(self) -> list[BotAction]:
        """Decide actions using neural network inference.

        Returns:
            List of BotAction tuples for keyboard/mouse inputs
        """
        ...


class NeuralNetBotRunner:
    """Runs a neural network bot connected to a game server.

    Handles integration between NeuralNetBot and GameClient,
    managing state updates and sending only changed inputs.
    """

    def __init__(
        self,
        client: GameClient,
        network: ActorCriticNetwork,
        config: NeuralNetBotConfig | None = None,
    ) -> None:
        """Initialize the bot runner.

        Args:
            client: GameClient for server communication
            network: Trained ActorCriticNetwork for inference
            config: Optional configuration for the neural net bot
        """
        ...

    async def on_game_state(self, state: GameState) -> None:
        """Handle incoming game state updates.

        Args:
            state: Current game state from the server
        """
        ...

    def reset(self) -> None:
        """Reset the bot runner state for a new game/episode."""
        ...
```

### Key Implementation Details

1. **Action Translation Logic**: The `Action` enum (0-26) maps to:
   - Actions 0-7: Movement key press/release (A, D, W, S)
   - Actions 8-23: Aim directions (16 buckets covering 0-2pi radians)
   - Actions 24-25: Shoot press/release
   - Action 26: No-op

2. **State Tracking**: Track previous movement key states, aim direction, and shooting state to only send changed inputs (similar to `RuleBasedBotRunner._previous_keyboard_actions`).

3. **Observation Building**: The `ObservationBuilder.build()` method requires the game state and player ID, returning a normalized float32 numpy array.

4. **Tensor Handling**: Convert numpy observation to PyTorch tensor, add batch dimension, run inference, then extract the action integer.

### Testing Strategy

**Unit Tests** (`bot2/tests/unit/bots/test_neural_net_bot.py`)
- [ ] Test `NeuralNetBot` initialization with network and config
- [ ] Test `NeuralNetBot.decide_actions()` returns valid `BotAction` tuples
- [ ] Test action translation for movement actions (press/release)
- [ ] Test action translation for aim actions (direction updates)
- [ ] Test action translation for shoot actions (mouse press/release)
- [ ] Test no-op action returns empty list
- [ ] Test `NeuralNetBotRunner` initialization
- [ ] Test `NeuralNetBotRunner.on_game_state()` sends correct inputs
- [ ] Test `NeuralNetBotRunner.reset()` clears state
- [ ] Test runner only sends changed inputs (deduplication)
- [ ] Test bot handles dead player state gracefully
- [ ] Test config defaults are applied correctly

**Edge Cases**
- [ ] Bot handles missing player in game state
- [ ] Bot handles no enemies in game state
- [ ] Runner handles client without player_id set
- [ ] Bot works with CUDA device when available

## Out of Scope

- WebSocket connection management (handled by TASK-002)
- Bot lifecycle management (handled by TASK-003)
- REST API endpoints (handled by TASK-004)
- Model loading from registry (handled by TASK-003)
- Integration with Go backend (handled by TASK-005)
- Training new models (existing training pipeline)

## References

- Epic: #95
- Design document: `squishy-drifting-otter.md`
- Related issue: #78 (Add bots to human game)
- `RuleBasedBot` implementation: `bot2/src/bot/bots/rule_based_bot.py`
- `ActorCriticNetwork` implementation: `bot2/src/bot/agent/network.py`
- `Action` enum: `bot2/src/bot/actions.py`
